pkg/apiserver/watch.go

```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+
+	t timeoutFactory
 }
 
-// ServeHTTP serves a series of JSON encoded events via straight HTTP with
-// Transfer-Encoding: chunked.
-func (self *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	loggedW := httplog.LogOf(req, w)
+// Serve serves a series of encoded events via HTTP with Transfer-Encoding: chunked
+// or over a websocket connection.
+func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	w = httplog.Unlogged(w)
-	timeoutCh, cleanup := self.t.TimeoutCh()
-	defer cleanup()
-	defer self.watching.Stop()
+
+	if wsstream.IsWebSocketRequest(req) {
+		w.Header().Set("Content-Type", s.mediaType)
+		websocket.Handler(s.HandleWS).ServeHTTP(w, req)
+		return
+	}
 
 	cn, ok := w.(http.CloseNotifier)
 	if !ok {
-		loggedW.Addf("unable to get CloseNotifier: %#v", w)
-		http.NotFound(w, req)
+		err := fmt.Errorf("unable to start watch - can't get http.CloseNotifier: %#v", w)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewInternalError(err), w, req)
 		return
 	}
 	flusher, ok := w.(http.Flusher)
 	if !ok {
-		loggedW.Addf("unable to get Flusher: %#v", w)
-		http.NotFound(w, req)
+		err := fmt.Errorf("unable to start watch - can't get http.Flusher: %#v", w)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewInternalError(err), w, req)
+		return
+	}
+
+	// get a framed encoder
+	f, ok := s.encoder.(streaming.Framer)
+	if !ok {
+		// programmer error
+		err := fmt.Errorf("no streaming support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
+		return
+	}
+	framer := f.NewFrameWriter(w)
+	if framer == nil {
+		// programmer error
+		err := fmt.Errorf("no stream framing support is available for media type %q", s.mediaType)
+		utilruntime.HandleError(err)
+		s.scope.err(errors.NewBadRequest(err.Error()), w, req)
 		return
 	}
+	e := streaming.NewEncoder(framer, s.encoder)
+
+	// ensure the connection times out
+	timeoutCh, cleanup := s.t.TimeoutCh()
+	defer cleanup()
+	defer s.watching.Stop()
+
+	// begin the stream
+	w.Header().Set("Content-Type", s.mediaType)
 	w.Header().Set("Transfer-Encoding", "chunked")
 	w.WriteHeader(http.StatusOK)
 	flusher.Flush()
-	// TODO: use arbitrary serialization on watch
-	encoder := watchjson.NewEncoder(w, self.encoder)
+
+	buf := &bytes.Buffer{}
 	for {
 		select {
 		case <-cn.CloseNotify():
 			return
 		case <-timeoutCh:
 			return
-		case event, ok := <-self.watching.ResultChan():
+		case event, ok := <-s.watching.ResultChan():
 			if !ok {
 				// End of results.
 				return
 			}
-			self.fixup(event.Object)
-			if err := encoder.Encode(&event); err != nil {
-				// Client disconnect.
+			obj := event.Object
+			s.fixup(obj)
+			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
+				// unexpected error
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				return
+			}
+			event.Object = &runtime.Unknown{
```

Comment:
No, the serialization of runtime.Unknown is unwrapped to send the actual
body for JSON and is the normal serialization for protobuf.

On Mon, Apr 4, 2016 at 1:01 PM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/apiserver/watch.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r58409606>
> :
>
> >  			if !ok {
> >  				// End of results.
> >  				return
> >  			}
> > -			self.fixup(event.Object)
> > -			if err := encoder.Encode(&event); err != nil {
> > -				// Client disconnect.
> > +			obj := event.Object
> > +			s.fixup(obj)
> > +			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
> > +				// unexpected error
> > +				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
> > +				return
> > +			}
> > +			event.Object = &runtime.Unknown{
>
> So that means that we are always sending "runtim.Unknown" as part of our
> watch.
> I'm not saying it's bad, but maybe we should document it somewhere.
>
> @lavalamp <https://github.com/lavalamp> - FYI ^^
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/26e6630171ddc8f3653bbcfa270efb87e1073818#r58409606>
>